### PR TITLE
PLANNER-2200: Update OptaPlanner version to 8.0.0-SNAPSHOT

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**
- Replace WannabeSolverManager with ScoreManager + SolverManager
- Replace OptaWeb's SolverStatus with OptaPlanner's SolverStatus
- Genericify Indictment usage
- Add scheduleUpdateOfRoster method to update shifts in a
  separate thread to not block Solver

### JIRA

https://issues.redhat.com/browse/PLANNER-2200

### Referenced pull requests

https://github.com/kiegroup/optaplanner/pull/1009 to know when it's safe to merge this PR.

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>full downstream build</b> please add comment: <b>Jenkins run fdb</b>
* for a <b>compile downstream build</b> please add comment: <b>Jenkins run cdb</b>
* for a <b>full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>
* for an <b>upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

File: pom.xml
Comment (triceo) [inline pom.xml:10]: You shouldn't be using KIE parent anymore. 8.0.0 does not need KIE parent.

**Review by triceo — CHANGES_REQUESTED** [review]
I like it, especially the removal of `WannabeSolverManager`! 
Just one comment inside.

File: pom.xml
Comment (Christopher-Chianelli) [reply pom.xml:10]: What should my parent pom be then (if any?)

Comment (triceo) [reply pom.xml:10]: OptaPlanner uses

    <parent>
      <groupId>org.jboss</groupId>
      <artifactId>jboss-parent</artifactId>
      <version>36</version>
      <relativePath/>
    </parent>

To get some basic commonality with middleware. Besides that, just import OptaPlanner BOM and you should be golden.

File: optaweb-employee-rostering-backend/pom.xml
Comment (triceo) [inline optaweb-employee-rostering-backend/pom.xml:33]: I believe that this is just `11`.

File: pom.xml
Comment (triceo) [inline pom.xml:254]: I think that the prod rule is "no fixed versions of anything".

File: optaweb-employee-rostering-backend/pom.xml
Comment (Christopher-Chianelli) [reply optaweb-employee-rostering-backend/pom.xml:33]: It is just 11, but it still built; odd

File: pom.xml
Comment (Christopher-Chianelli) [reply pom.xml:254]: Copied from kie-parent (needed to set a lot of dependency versions to make it build again)

Comment (triceo) [reply pom.xml:254]: Right. My point is this should also be specified as a property.
Alternatively we could use the `optaplanner-build-parent`, but I'm not sure if that is proper. We'd have to ask Geoffrey what his idea for that is.

File: optaweb-employee-rostering-backend/pom.xml
Comment (triceo) [reply optaweb-employee-rostering-backend/pom.xml:33]: Perhaps it's not being used anywhere, then?

Comment (Christopher-Chianelli) [reply optaweb-employee-rostering-backend/pom.xml:33]: apparently `java.version` is only if you use the spring boot starter parent (we don't). So we should use `<maven.compiler.release>11</maven.compiler.release>` instead

**Review by triceo — APPROVED** [review]
LGTM

**Comment by triceo** [comment]
You should probably update CI configs - remove Java 8, change 14 to 15.

**Comment by triceo** [comment]
The Jenkins failure will go away with https://github.com/kiegroup/optaweb-employee-rostering/pull/509

File: optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/roster/RosterService.java
Comment (yurloc) [inline optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/roster/RosterService.java:418]: `Optional.get()` without `isPresent()` check. A simple way to avoid it is to use `forEach()` as the terminal operation and unassigning the employee there.

Comment (yurloc) [reply optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/roster/RosterService.java:418]: Abbreviations as variable names should be avoided. So `constraintMatch` instead of `cm`.

Using a full name that has a meaning avoids the mental mapping overhead. When reading `cm.getJustificationList()` I have to go back and guess that elements of `getConstraintMatchSet()` are `ConstraintMatch`es. This is a simple example but it illustrates the mental mapping problem.

The Clean Code book from Robert C. Martin has a nice chapter on naming.

**Review by yurloc — CHANGES_REQUESTED** [review]
A few code style and dependency issues, otherwise looks good.

File: optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/roster/RosterService.java
Comment (Christopher-Chianelli) [reply optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/roster/RosterService.java:414]: Can't find duplicate, but it was already a constant: `RosterConstraintConfiguration.CONSTRAINT_UNAVAILABLE_TIME_SLOT_FOR_AN_EMPLOYEE`; replaced it here.

**Comment by Christopher-Chianelli** [comment]
Do not merge; local test failures (Drools causing a NPE in tests)

File: pom.xml
Comment (yurloc) [inline pom.xml:65]: Not used.

File: optaweb-employee-rostering-standalone/pom.xml
Comment (yurloc) [inline optaweb-employee-rostering-standalone/pom.xml:54]: This shouldn't be necessary if you bind the `analyze-only` execution to `none` phase.

File: pom.xml
Comment (yurloc) [inline pom.xml:47]: Please do the same for `optaweb-vehicle-routing`.

**Review by yurloc — CHANGES_REQUESTED** [review]
Spring Boot starter dependencies should be used transitively. Otherwise looks good.

**Comment by triceo** [comment]
I believe that the Drools fails should now be gone. Jenkins please retest this.

**Comment by triceo** [comment]
No, my bad, https://github.com/kiegroup/kogito-runtimes/pull/858 not yet merged.

**Comment by Christopher-Chianelli** [comment]
@triceo Standalone broken, but spring-boot:run works. No NPE in test case, but the test case is saying the constraint is not triggering (i.e. test still failing, but not with a NPE).

Exception from standalone seems to be related to the generated files from constraint streams/drools, unable to find certain classes.

Example message:
```
Message [id=75, level=ERROR, path=src/main/java/org/drools/project/model/ProjectModel_org_default_artifact_1_0_0.java, line=4, column=27
   text=cannot access org.kie.api.builder.ReleaseId
  bad class file: file:/mnt/external-drive/code/kiegroup/optaplanner/employee-rostering/optaweb-employee-rostering-standalone/target/optaweb-employee-rostering-standalone-8.0.0-SNAPSHOT-exec.jar!/org/kie/api/builder/ReleaseId.class
    unable to access file: java.io.FileNotFoundException: JAR entry org/kie/api/builder/ReleaseId.class not found in /mnt/external-drive/code/kiegroup/optaplanner/employee-rostering/optaweb-employee-rostering-standalone/target/optaweb-employee-rostering-standalone-8.0.0-SNAPSHOT-exec.jar
    Please remove or make sure it appears in the correct subdirectory of the classpath.]
Message [id=76, level=ERROR, path=src/main/java/org/drools/project/model/ProjectModel_org_default_artifact_1_0_0.java, line=5, column=36
   text=cannot access org.drools.compiler.kproject.ReleaseIdImpl
  bad class file: file:/mnt/external-drive/code/kiegroup/optaplanner/employee-rostering/optaweb-employee-rostering-standalone/target/optaweb-employee-rostering-standalone-8.0.0-SNAPSHOT-exec.jar!/org/drools/compiler/kproject/ReleaseIdImpl.class
    unable to access file: java.io.FileNotFoundException: JAR entry org/drools/compiler/kproject/ReleaseIdImpl.class not found in /mnt/external-drive/code/kiegroup/optaplanner/employee-rostering/optaweb-employee-rostering-standalone/target/optaweb-employee-rostering-standalone-8.0.0-SNAPSHOT-exec.jar
    Please remove or make sure it appears in the correct subdirectory of the classpath.]
Message [id=77, level=ERROR, path=src/main/java/org/drools/project/model/ProjectModel_org_default_artifact_1_0_0.java, line=7, column=89
   text=cannot access org.drools.modelcompiler.CanonicalKieModuleModel
  bad class file: file:/mnt/external-drive/code/kiegroup/optaplanner/employee-rostering/optaweb-employee-rostering-standalone/target/optaweb-employee-rostering-standalone-8.0.0-SNAPSHOT-exec.jar!/org/drools/modelcompiler/CanonicalKieModuleModel.class
    unable to access file: java.io.FileNotFoundException: JAR entry org/drools/modelcompiler/CanonicalKieModuleModel.class not found in /mnt/external-drive/code/kiegroup/optaplanner/employee-rostering/optaweb-employee-rostering-standalone/target/optaweb-employee-rostering-standalone-8.0.0-SNAPSHOT-exec.jar
    Please remove or make sure it appears in the correct subdirectory of the classpath.]
Message [id=78, level=ERROR, path=src/main/java/org/drools/project/model/ProjectModel_org_default_artifact_1_0_0.java, line=15, column=27
   text=cannot find symbol
  symbol:   class Model
  location: class org.drools.project.model.ProjectModel_org_default_artifact_1_0_0]
Message [id=79, level=ERROR, path=src/main/java/org/drools/project/model/ProjectModel_org_default_artifact_1_0_0.java, line=19, column=27
   text=cannot find symbol
  symbol:   class Model
  location: class org.drools.project.model.ProjectModel_org_default_artifact_1_0_0]
Message [id=80, level=ERROR, path=src/main/java/org/drools/project/model/ProjectModel_org_default_artifact_1_0_0.java, line=27, column=12
   text=cannot find symbol
  symbol:   class ReleaseId
  location: class org.drools.project.model.ProjectModel_org_default_artifact_1_0_0]
Message [id=81, level=ERROR, path=src/main/java/org/drools/project/model/ProjectModel_org_default_artifact_1_0_0.java, line=32, column=33
   text=cannot access org.kie.api.builder.model.KieModuleModel
  bad class file: file:/mnt/external-drive/code/kiegroup/optaplanner/employee-rostering/optaweb-employee-rostering-standalone/target/optaweb-employee-rostering-standalone-8.0.0-SNAPSHOT-exec.jar!/org/kie/api/builder/model/KieModuleModel.class
    unable to access file: java.io.FileNotFoundException: JAR entry org/kie/api/builder/model/KieModuleModel.class not found in /mnt/external-drive/code/kiegroup/optaplanner/employee-rostering/optaweb-employee-rostering-standalone/target/optaweb-employee-rostering-standalone-8.0.0-SNAPSHOT-exec.jar
    Please remove or make sure it appears in the correct subdirectory of the classpath.]
Message [id=82, level=ERROR, path=src/main/java/org/drools/project/model/ProjectModel_org_default_artifact_1_0_0.java, line=0, column=0
   text=Java source of src/main/java/org/drools/project/model/ProjectModel_org_default_artifact_1_0_0.java in error:
```

**Comment by triceo** [comment]
@Christopher-Chianelli This has got to be on your side now. Check the dep tree, compare it against dep tree of Optaplanner. They should be identical on the KIE/Kogito front.

**Comment by triceo** [comment]
Jenkins retest this.

**Comment by triceo** [comment]
OK, it seems we do still have a Drools failure in there. 
@Christopher-Chianelli or @yurloc, can you please put together a simplified reproducer?

**Comment by yurloc** [comment]
Looking.

**Comment by yurloc** [comment]
[ERROR] Failures: 
[ERROR]   DroolsSolverTest>AbstractSolverTest.testContractConstraints:250->AbstractSolverTest.testContractConstraint:236 expected: <-60> but was: <0>

I have isolated a reproducer for this and reported it as https://issues.redhat.com/browse/DROOLS-5789.

I don't know whether this Drools bug could be related to the problem with standalone JAR. I don't think so. We should probably continue investigating that.

**Comment by Christopher-Chianelli** [comment]
The error is definitely related to the usage of scoreDRL; the executable jar works for me when I disable `scoreDrl` and enabled the `constraintProvider`.

**Comment by triceo** [comment]
I have just seen a green local build of this branch after applying this PR first:
https://github.com/kiegroup/drools/pull/3211

The end is near. I will let you know once that PR is merged and Kogito is upgraded to use it.

**Comment by triceo** [comment]
Furthermore, I can confirm the issue with the standalone JAR.

**Comment by Christopher-Chianelli** [comment]
@triceo confirm it is removed after the drools PR or confirm it is still there with the drools PR?

**Comment by triceo** [comment]
The issue is still there. It looks completely unrelated to the Drools PR.

File: optaweb-employee-rostering-standalone/pom.xml
Comment (triceo) [inline optaweb-employee-rostering-standalone/pom.xml:83]: Formatting?

Comment (Christopher-Chianelli) [reply optaweb-employee-rostering-standalone/pom.xml]: Fixed; thanks! (amended commit)

File: optaweb-employee-rostering-standalone/pom.xml
Comment (yurloc) [inline optaweb-employee-rostering-standalone/pom.xml:107]: Wrong indentation.

**Comment by yurloc** [comment]
Although the shade plugin approach the classloading issue I think it does it in a wrong way. Now I can see this:

2020-11-04 19:09:01.674  INFO 277863 --- [  restartedMain] o.s.b.d.a.OptionalLiveReloadServer       : **LiveReload server is running on port 35729**
2020-11-04 19:09:01.990  INFO 277863 --- [  restartedMain] o.h.h.i.QueryTranslatorFactoryInitiator  : HHH000397: Using ASTQueryTranslatorFactory
WARNING: An illegal reflective access operation has occurred
WARNING: **Illegal reflective access by org.springframework.cglib.core.ReflectUtils** (file:/home/jlocker/src/github.com/kiegroup/optaweb-employee-rostering/optaweb-employee-rostering-standalone/target/optaweb-employee-rostering-standalone-8.0.0-SNAPSHOT-exec.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of org.springframework.cglib.core.ReflectUtils
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
2020-11-04 19:09:02.518  INFO 277863 --- [  restartedMain] o.o.s.b.a.OptaPlannerAutoConfiguration   : THIS: jdk.internal.loader.ClassLoaders$AppClassLoader@5bc2b487->jdk.internal.loader.ClassLoaders$PlatformClassLoader@77ec153e
2020-11-04 19:09:02.518  INFO 277863 --- [  restartedMain] o.o.s.b.a.OptaPlannerAutoConfiguration   : BEAN: org.springframework.boot.devtools.restart.classloader.RestartClassLoader@2802a4e5->jdk.internal.loader.ClassLoaders$AppClassLoader@5bc2b487->jdk.internal.loader.ClassLoaders$PlatformClassLoader@77ec153e

The last two lines are debug output I added to optaplanner spring boot starter to see which classloaders are used. OptaPlannerAutoConfiguration is loaded by the app classloader but the bean classloader that it uses to create solver factory is the devtools restart classloader.

**Comment by Christopher-Chianelli** [comment]
The URL Classloader provided by Swing seems to be unable to find classes in the libs folder of the executable Jar. We can add a workaround comment to it; the release is coming up and I have a feeling this will be a very non-trivial fix.

**Comment by Christopher-Chianelli** [comment]
The same error (unable to load class files) is present in optaweb-vehicle-routing if one uses the old Drools file (commenting out the last constraint as it references a method that was removed).

**Comment by triceo** [comment]
https://github.com/kiegroup/optaplanner/pull/1009
When this is merged, this PR can be merged as well.

**Comment by Christopher-Chianelli** [comment]
Once https://github.com/kiegroup/drools/pull/3220 is merged and kogito is updated, this can be merged (last commit reverted as a fix to the class loader issue in drools was found)

**Review by yurloc — APPROVED** [review]

**Comment by triceo** [comment]
Jenkins retest this.

**Comment by triceo** [comment]
The Sonar warning deserves to be looked at.

**Comment by Christopher-Chianelli** [comment]
@dupliaka I fixed https://issues.redhat.com/browse/PLANNER-2189 in the latest commit (small change in frontend code)

**Comment by triceo** [comment]
How can the Build Chain CI be green if a test failed there?

    Error:    RosterRestControllerTest.testGetShiftRosterAsExcel:573 [Response should be successful for one spot] 
    Expecting:
     <500 INTERNAL_SERVER_ERROR>
    to be equal to:
     <200 OK>
    but was not.
    [INFO] 
    Error:  Tests run: 136, Failures: 1, Errors: 0, Skipped: 2

**Comment by Christopher-Chianelli** [comment]
@triceo the error did not reproduce locally or on Jenkins. So why does it happen on build chain (and why did it still pass the build chain?)

**Comment by Christopher-Chianelli** [comment]
@triceo I see this in the maven command: `-Dmaven.test.failure.ignore=true `

**Comment by Christopher-Chianelli** [comment]
It did not happen in the CI GitHub actions... did something happen that broke excel functionality higher in the build chain?

**Comment by triceo** [comment]
@rsynek Are we aware of this? Tests failing the Build Chain being disabled?
@Christopher-Chianelli If it fails on Build Chain and not on "OS x Java", could it be a flake?

**Comment by Christopher-Chianelli** [comment]
@triceo it not failing locally, on Jenkins, or on CI. But nothing in the test in question could be flakey (unless you are running it on a machine with barely enough memory to even run a JVM, in which case everything would be flakey). None of the method called rely on any timing; only loading from the database.

The test:
```
    @Test
    public void testGetShiftRosterAsExcel() {
        createTestRoster();

        LocalDate startDate = LocalDate.of(2000, 1, 1);
        LocalDate endDate = LocalDate.of(2000, 1, 2);

        List<Spot> requestSpotList = spotList.subList(0, 1);
        ResponseEntity<byte[]> response = getShiftRosterAsExcel(requestSpotList, startDate, endDate);

        // This test fail
        assertThat(response.getStatusCode()).as("Response should be successful for one spot").isEqualTo(HttpStatus.OK);
        try (Workbook workbook = new XSSFWorkbook(new ByteArrayInputStream(response.getBody()))) {
            assertThat(workbook.getNumberOfSheets()).as("There should only be one sheet in the file").isEqualTo(1);
            Sheet spotSheet = workbook.getSheet(requestSpotList.get(0).getName());
            assertSheetForSpotIsCorrect(spotSheet, requestSpotList.get(0), startDate, endDate);
        } catch (IOException e) {
            fail("Unable to read response as an Excel file");
        }

        requestSpotList = spotList;
        response = getShiftRosterAsExcel(requestSpotList, startDate, endDate);

        assertThat(response.getStatusCode()).as("Response should be successful for two spots").isEqualTo(HttpStatus.OK);
        try (Workbook workbook = new XSSFWorkbook(new ByteArrayInputStream(response.getBody()))) {
            assertThat(workbook.getNumberOfSheets()).as("There should only be two sheets in the file").isEqualTo(2);
            Sheet spotSheet = workbook.getSheet(requestSpotList.get(0).getName());
            assertSheetForSpotIsCorrect(spotSheet, requestSpotList.get(0), startDate, endDate);
            spotSheet = workbook.getSheet(requestSpotList.get(1).getName());
            assertSheetForSpotIsCorrect(spotSheet, requestSpotList.get(1), startDate, endDate);
        } catch (IOException e) {
            fail("Unable to read response as an Excel file");
        }

        Spot badSpot = new Spot(TENANT_ID, "Bad", Collections.emptySet());
        badSpot.setId(-1L);

        requestSpotList = Arrays.asList(badSpot);
        response = getShiftRosterAsExcel(requestSpotList, startDate, endDate);

        assertThat(response.getStatusCode()).as("Response should fail with BAD_REQUEST for non-existing spot in tenant")
                .isEqualTo(HttpStatus.BAD_REQUEST);
    }
```

**Comment by Christopher-Chianelli** [comment]
I did notice `@Deprecation` warnings in the logs for classes used in processing excel. Might be related.

**Comment by rsynek** [comment]
Hi @Ginxo, may I ask why do we use `-Dmaven.test.failure.ignore=true` in https://github.com/kiegroup/droolsjbpm-build-bootstrap/blob/master/.ci/pull-request-config.yaml#L7? The PR check is green despite some test failed.

**Comment by rsynek** [comment]
Merging despite the failure discovered in the Build Chain (see the discussion above).

**Pull request merged**
